File: objc/Visitor/Visitor/Library.m
#import "Library.h"

@implementation Library

-(instancetype)initWith:(NSString*)name {
    if(self = [super init]) {
        _name = name;
        _bookList = [[NSMutableArray alloc] init];
    }
    return self;
}

-(void)addBook:(Book*)book {
    [_bookList addObject:book];
}

-(void)accept:(id<LibraryVisitor>)visitor {
    for(Book* book in _bookList)
        [visitor visitBook:book];
    [visitor visitLibrary:self];
}

@end

File: swift/Visitor/Visitor/Library.swift
import Foundation

class Library : Visitable {
    var name: String?
    var bookList: [Book] = [Book]()
    
    init(_ name: String) {
        self.name = name
    }
    
    func addBook(_ book: Book) -> Void {
        bookList.append(book)
    }
    
    func accept(_ visitor: LibraryVisitor) {
        for book in bookList {
            visitor.visitBook(book)
        }
        visitor.visitLibrary(self)
    }
}
